lib/streamlit/cli.py

```diff
@@ -186,23 +199,12 @@ def main_run(file_or_url, args=None, **kwargs):
     _apply_config_options_from_cli(kwargs)
 
     if url(file_or_url):
-        import tempfile
-        import requests
-
-        with tempfile.NamedTemporaryFile() as fp:
-            try:
-                resp = requests.get(file_or_url)
-                resp.raise_for_status()
-                fp.write(resp.content)
-                # flush since we are reading the file within the with block
-                fp.flush()
-            except requests.exceptions.RequestException as e:
-                raise click.BadParameter(
-                    ("Unable to fetch {}.\n{}".format(file_or_url, e))
-                )
-            # this is called within the with block to make sure the temp file
-            # is not deleted
-            _main_run(fp.name, args)
+        from streamlit.temporary_directory import TemporaryDirectory
+
+        with TemporaryDirectory() as temp_dir:
+            script_path = os.path.join(temp_dir, os.path.basename(file_or_url))
```

Comment:
Don't use basename with URLs. It's meant to be used with filesystem paths, so on Windows it will use backslashes. Also, does it doesn't work when the URL ends in a slash.

Use something like `urllib.parse.urlparse()` instead.

Reply:
Also, to make sure this doesn't break in the future, can you add a test for `main_run`?

If tests don't work with `click`, you can always pull the entire body of `main_run` into another function called `_main_run_real`:

```
@stuff
@stuff
def main_run(...):
  main_run_real(...)
```

Reply:
We have tests for this. I definitely had to update them :-)